include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2021 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+#define CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+
+#ifndef SPINLOCK_HINT
+#ifdef CPP_UTILITY_HAS_SPINLOCK_HINT
+#include <xmmintrin.h>
+#define SPINLOCK_HINT _mm_pause();  // NOLINT
+#else
+#define SPINLOCK_HINT /* do nothing */
+#endif
+#endif
+
+#include <xmmintrin.h>
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+#include <utility>
+
+namespace dbgroup::lock
+{
+class OptimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  constexpr OptimisticLock() = default;
+
+  OptimisticLock(const OptimisticLock&) = delete;
+  OptimisticLock(OptimisticLock&&) = delete;
+
+  auto operator=(const OptimisticLock&) -> OptimisticLock& = delete;
+  auto operator=(OptimisticLock&&) -> OptimisticLock& = delete;
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~OptimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+
+  /**
+   * @brief Get the Version object
+   *
+   * @return std::pair<uint64_t, bool>
+   */
+  [[nodiscard]] auto
+  GetVersion()  //
+      -> std::pair<uint64_t, bool>
+  {
+    version_ = lock_.load(std::memory_order_relaxed);
+    if (version_ & (2UL) == 0b10) return std::make_pair(version_, false);
+    return std::make_pair(version_, true);
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
+   */
+  auto
+  CheckVersion()  //
+      -> bool
+  {
+    std::pair<uint64_t, bool> current_version_ = GetVersion();
+    if (!GetVersion().second) return false;
+    if (current_version_.first == version_) return true;
+    return false;
+  }
```

Comment:
バージョンをもらう前提にするので，これだけでOKですね．

```suggestion
  auto
  CheckVersion(const uint64_t expected) const  //
      -> bool
  {
    const auto desired = lock_.load(std::memory_order_relaxed) & kXLockMask;
    return expected == desired;
  }
```

Reply:
直感的には`std::memory_order_acquire`な気がしたのですが，`std::memory_order_relaxed`で十分な理由は何でしょうか？

Reply:
`GetVersion`側でacquireフェンスを張っているからですね．バージョンチェックを行う際は`GetVersion`後に他のスレッドが何も編集していないことを期待するので，誰もreleaseフェンスを張っていないという想定になります．

Reply:
理解しました！ありがとうございます！